packages/next/server/app-render.tsx

```diff
@@ -305,6 +301,12 @@ export async function renderToHTML(
   renderOpts: RenderOpts,
   isPagesDir: boolean
 ): Promise<RenderResult | null> {
+  // this needs to be required lazily so that `next-server` can set
+  // the env before we require
+  const ReactDOMServer = process.env.__NEXT_REACT_ROOT
```

Comment:
we can use the `shouldUseReactRoot` flag instead of process env to check which react-dom to use which doesn't require any preprocess
```js
import { shouldUseReactRoot } from './utils'

const ReactDOMServer = shouldUseReactRoot
  ? require('react-dom/server.browser')
  : require('react-dom/server')
```

Reply:
Good point, updated in https://github.com/vercel/next.js/pull/38420/commits/11cc8e7ed63edb1d2df8da01af2e51a2c10f9e5e